### ISSUE TITLE
PShapeOpenGL indexed set() methods trigger inefficient re-tessellation

### ISSUE THREAD

**Issue by mcslee** [processing/processing#5071]
The PShapeOpenGL API offers support for updating the tint/fill/stroke values for a single vertex via the setTint(int index, int tint) API:
https://github.com/processing/processing/blob/master/core/src/processing/opengl/PShapeOpenGL.java#L1997

In order to refresh the color VBO, this method invokes a call to markForTessellation():
https://github.com/processing/processing/blob/master/core/src/processing/opengl/PShapeOpenGL.java#L2005

This is overkill - markForTesselation() triggers re-tessellation of the entire object's geometry. But really, we *only* need to update the color buffer data. Note that the implementation of the non-indexed setTint(int tint) API does this efficiently, by invoking root.setModifiedPolyColors()
https://github.com/processing/processing/blob/master/core/src/processing/opengl/PShapeOpenGL.java#L1982 - this appropriately triggers updating the color buffer to the GPU, but not the rest of the unmodified buffers.

A common use for PShape which I have run into is to animate a fixed geometry - say for instance, a huge cloud of points, representing LEDs that don't change position but that change color every frame. To get good efficiency here, we only want to re-upload the color buffer data to the GPU on each rendering pass.

I have hacked up my own PShapeOpenGL subclass which accomplishes this by updating the buffer directly without using the setTint() API, but it's a hacky solution. For reference, that code can be seen here: https://github.com/treeoftenere/Tenere/commit/892378dee900dd2b8e262b7aeace849e461ef119#diff-82c4eab403617e7da8960538139bedd8R221

It would be great if the following PShapeOpenGL APIs used the root.setModified*() family of methods for a more efficient buffer update. The mechanism is already in place to do these selective updates. This would yield a major performance improvement.

Applies to:
PShapeOpenGL.setTint(int index, int tint)
PShapeOpenGL.setFill(int index, int fill)
PShapeOpenGL.setStroke(int index, int stroke)

It's possible that I'm missing something here. I did notice the comment at line 1654:
https://github.com/processing/processing/blob/master/core/src/processing/opengl/PShapeOpenGL.java#L1654

But I don't think that comment applies to modifications of non-geometric vertex data.

**Comment by codeanticode** [comment]
@mcslee Many  thanks for your comments on PShapeOpenGL, I agree that its architecture is not particularly optimized. This relates to #1623, which was pushed for later at the time. Maybe it can be taken care of now, I'm curious about your modifications, will take a look as soon as I get the chance.

**Comment by mcslee** [comment]
@codeanticode Thank you for reading and considering!

My code example is probably not the best guide here, as I am totally abusing the API and reaching into the internals with my own bindBuffer() and bufferData() calls (bypassing the internal geometry objects). It's a bit of a nasty hack, but it at least demonstrates what I would like to be able to accomplish from an efficiency POV.

Cheers!

**Comment by codeanticode** [comment]
Ok, the problem to support color updates generally, is that you need to know what vertices in the tessellated geometry correspond to each input vertex, and that correspondence may not be one-to-one, specially in the case of poly shapes that use curve or Bezier vertices. It would not impossible to do, but tricky. Perhaps one could address special cases such as triangles, quads, etc. where the correspondence is straightforward.

**Comment by mcslee** [comment]
Makes sense. Yeah, my use case was simple geometry, QUADS specifically but POINTS and TRIANGLES would be similarly simple. I can see the risk of this turning into spaghetti-code with too much special-casing by geometry type (though there already a fair bit of that in this file, e.g. if family == GROUP etc. type of stuff). It sort of seems like the more elegant solution would be to have a class hierarchy of PShape subclasses for the different geometries, but that is obviously a much larger refactor.

Sub-classing for higher-performance is okay for my use case. One issue I ran into sub-classing though: I had to use my own IntBuffer objects, since a subclass that's not part of the processing.core.opengl package cannot access the TessGeometry class, as it is protected static. (Same applies to InGeometry and Tesselator, though I don't think I needed those).

It would have been nice if my PShapeOpenGL subclass were able to access tessGeo.polyColors. Then I could have simply updated those values manually and used the root.setModifiedPolyColors() method to trigger the buffer being updated on the next draw loop. In fact, setModifiedPolyColors() really already contains most of this "smart partial redraw" logic...

Could perhaps consider declaring TessGeometry public? Though I understand if that's a yucky "exposing too much guts" for the general case. And I know subclassing Processing's implementation classes is not something that would ever be encouraged or supported. But PShapeOpenGL is just *so close* to meeting this need.
